rust_snuba/src/processors/generic_metrics.rs

```diff
@@ -78,6 +81,60 @@ enum MetricValue {
     },
 }
 
+#[derive(Debug, Deserialize)]
+#[serde(tag = "format", rename_all = "lowercase")]
+enum EncodedSeries<T> {
+    Array { data: Vec<T> },
+}
+
+impl<T> EncodedSeries<T> {
+    fn into_vec(self) -> Vec<T> {
+        match self {
+            EncodedSeries::Array { data } => data,
+        }
+    }
+}
+
+fn encoded_series_deserializer<'de, T, D>(deserializer: D) -> Result<EncodedSeries<T>, D::Error>
+where
+    T: Deserialize<'de>,
+    D: Deserializer<'de>,
+{
+    struct Visitor<U> {
+        phantom: PhantomData<U>,
+    }
+
+    impl<'de, U> serde::de::Visitor<'de> for Visitor<U>
+    where
+        U: Deserialize<'de>,
+    {
+        type Value = EncodedSeries<U>;
+
+        fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+            formatter.write_str("expected either old or new distribution value format")
+        }
+
+        fn visit_seq<A>(self, seq: A) -> Result<Self::Value, A::Error>
+        where
+            A: serde::de::SeqAccess<'de>,
+        {
+            let data = Vec::<U>::deserialize(SeqAccessDeserializer::new(seq))?;
+            Ok(EncodedSeries::Array { data })
+        }
+
+        fn visit_map<A>(self, map: A) -> Result<Self::Value, A::Error>
+        where
+            A: serde::de::MapAccess<'de>,
+        {
+            EncodedSeries::deserialize(MapAccessDeserializer::new(map))
+        }
+    }
+
+    deserializer.deserialize_any(Visitor {
+        phantom: PhantomData::<T>,
```

Comment:
```suggestion
        phantom: PhantomData,
```

You don't need to specify the type here, Rust can infer it form the return type.

With the Visitor as a unit struct, it becomes `Visitor(PhantomData)`, quite short.

Reply:
That's really cool, thanks